ENH: Add shape to *_like() array creation

pentschev opened the pull request:
This solves https://github.com/numpy/numpy/issues/13043.

@shoyer @mrocklin @hameerabbasi

pentschev committed:
ENH: Added shape argument to *_like() array creation functions

pentschev committed:
ENH: C backend adjustments for shape argument on *_like()

pentschev committed:
TST: Added test for shape argument in *_like() functions

pentschev commented:
Please note I've optimistically added the `versionadded` tag to the documentation as version 1.17. :)

rgommers commented on doc/source/reference/c-api.array.rst:
```diff
@@ -243,7 +243,7 @@ From scratch
 .. c:function:: PyObject* PyArray_NewLikeArray( \
         PyArrayObject* prototype, NPY_ORDER order, PyArray_Descr* descr, \
-        int subok)
+        int ndim, npy_intp* dims, int subok)
```
looks like this breaks backwards compatibility

shoyer replied:
Indeed -- unfortunately we can't add add parameters to functions in NumPy's C API without breaking API compatibility.

I think we need to make a private internal function to handle `shape` on the C side, which could be used by `empty_like()` and `PyArray_NewLikeArray` internally.

pentschev replied:
I see, thanks for pointing that out @rgommers.

@shoyer could you clarify how we would use such a function from `PyArray_NewLikeArray()`? I can see how I could have a completely new function `PyArray_NewLikeArrayWithShape()` being called from `array_empty_like()`, but that would incur on a lot of code duplication. I can't see how we would handle `PyArray_NewLikeArray()` if it doesn't know about the new shape we're trying to use, other than modifying `*prototype`'s shape before and restore it afterwards, but that feels too hacky.

pentschev replied:
Ah, ok, one alternative would be to move the implementation to a new function, and keep `PyArray_NewLikeArray()` only as a wrapper for backwards compatibility, is that what you mean @shoyer? I think this new function didn't even need to be private, it could be the new `PyArray_NewLikeArrayWithShape()` I proposed.

rgommers replied:
indeed, move the implementation to your new function and keep `PyArray_NewLikeArray` as a wrapper for backwards compat

rgommers commented:
adding `shape=None` to `*_like` seems reasonable to me, +1

pentschev committed:
ENH: Added PyArray_NewLikeArrayWithShape()

This change maintains backwards compatibility, rather than passing new
arguments to PyArray_NewLikeArray().

pentschev commented:
Alright, I think with the changes I just pushed, we're not breaking the API. Should I also update `cversions.txt` and `setup_common.py`, since now the API version needs to be incremented?

charris commented:
> Should I also update `cversions.txt` and `setup_common.py`

Yes.

eric-wieser commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1185,26 +1185,32 @@ PyArray_NewFromDescrAndBase(
     PyObject *ret = NULL;
-    int ndim = PyArray_NDIM(prototype);
+
+    if (ndim == 0 || dims == NULL) {
```
Ndim=0 should not be a special case, the caller might be asking for a 0d array

pentschev replied:
Yes, you're right, I'll have it fixed. Thanks for pointing it out.

eric-wieser commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1246,11 +1252,10 @@ PyArray_NewLikeArray(PyArrayObject *prototype, NPY_ORDER order,
 
-        PyArray_CreateSortedStridePerm(PyArray_NDIM(prototype),
+        PyArray_CreateSortedStridePerm(ndim,
                                         PyArray_STRIDES(prototype),
```
This is still broken - the strides array is off a different length to ndim. This isn't just an implementation bug, this is a design problem. What does it mean to say "create an Nd array matching the order of this Md one" when M != N

My hunch would be to say "it's meaningless, require the caller to insert dimensions to match by erroring"

pentschev replied:
I guess you're talking exclusively about the new case, creating an array with a different shape, right? In that case you're right, looking at the strides of the original array makes no sense, I will change it such that arrays with new shapes will always fall into the simpler `order != NPY_KEEPORDER` condition, where strides are ignored.

pentschev committed:
BUG: Fix for PyArray_NewLikeArrayWithShape strides and ndim == 0

Arrays created with new shapes should not take into consideration the original
array's stride, and ndim == 0 should not be a case to ignore a new shape, as
the caller may request a 0d array.

pentschev committed:
REL: Updates for C-API, version 1.17.x

* Add comments to cversions.txt (new PyArray_NewLikeArrayWithShape function)
* Increment C_API_VERSION to 1.17 in setup_common.py

mrocklin commented:
It looks like this is ready for another round of review

mattip commented:
Do we need to expose this in the public numpy C API, or can it be done only at the python level? It would be nice to keep the C API to the minimum required, and this seems to be yet-another-wrapper around `PyArray_NewFromDescr`

pentschev commented:
@mattip we need to expose the new `PyArray_NewLikeArrayWithShape()`, since the implementation was moved in there from `PyArray_NewLikeArray()`, we could now not expose the latter anymore. Is that what you mean?

mattip commented:
We cannot currently remove already published API functions. Is there a need to expose the new function at the C-API level? Do we expect C-extension writers to use `PyArray_NewLikeArrayWithShape` from C or can we get by with the python-level additions at this point? I would prefer not to expand the already massive C-API with another function.

hameerabbasi commented:
@pentschev What @mattip means is that we should simply remove the function from the header and make it internal to NumPy. That way we avoid changing the C API.

pentschev commented:
> We cannot currently remove already published API functions. Is there a need to expose the new function at the C-API level? Do we expect C-extension writers to use `PyArray_NewLikeArrayWithShape` from C or can we get by with the python-level additions at this point? I would prefer not to expand the already massive C-API with another function.

Not at the moment, but what is normally the C-API policy of NumPy? Should it avoid exposing new functionality if there's no known immediate use case? Sorry if this is a question that has been answered before.

pentschev commented:
> @pentschev What @mattip means is that we should simply remove the function from the header and make it internal to NumPy. That way we avoid changing the C API.

Sorry, I didn't see that comment before. I understand it, and that's doable. My previous question is a general one, should we _always_ prefer not exposing C-API, or is there a general rule when we should choose to expose it?

charris commented:
I don't recall that we have an explicit policy apart from not removing functions from the API, although they may be deprecated. Generally, new functions are added along with new basic functionality, particularly when the new functionality starts on the C side. The addition should also be discussed on the mailing list, and these days we might want to have an NEP as well. Needs some discussion.

pentschev commented:
@mrocklin I could use a suggestion from you here, do you think for our use case with Dask and libraries like CuPy and Sparse, do you see us needing to pass `shape` to `*_like()` functions in the foreseeable future via the C-API? If not, maybe we can skip NumPy's C-API change.

mrocklin commented:
From an applications perspective I do not anticipate the libraries above needing to use the C-API.  If dropping that accelerates things, then I'm for dropping.

pentschev committed:
Revert "REL: Updates for C-API, version 1.17.x"

This reverts commit 807f512ebeb7797ad374d845e41015948afcc708.

pentschev committed:
Revert exposing PyArray_NewLikeArrayWithShape on C-API

pentschev commented:
Ok, now there's no C-API changes anymore. It's ready for review again.

mattip commented on numpy/core/multiarray.py:
```diff
@@ -98,6 +98,10 @@ def empty_like(prototype, dtype=None, order=None, subok=None):
+        .. versionadded:: 1.17.0
+    shape : int or sequence of ints, optional.
+        Overrides the shape of the result.
+
```
The versionadded goes below the attribute, [here](https://6247-908607-gh.circle-artifacts.com/0/home/circleci/repo/doc/build/html/reference/generated/numpy.empty_like.html?highlight=empty_like) is how this is rendered. The *New in version 1.17.0.* is showing in the previous parameter

pentschev replied:
I see it now, the format of linespaces is a bit misleading there. I'll fix it now.

mattip commented on numpy/core/numeric.py:
```diff
@@ -136,6 +136,10 @@ def zeros_like(a, dtype=None, order='K', subok=True):
+        .. versionadded:: 1.17.0
+    shape : int or sequence of ints, optional.
+        Overrides the shape of the result.
+
```
versionadded below the parameter

mattip commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -1846,23 +1846,25 @@ static PyObject *
 {
 
-    static char *kwlist[] = {"prototype","dtype","order","subok",NULL};
+    static char *kwlist[] = {"prototype","dtype","order","subok","shape",NULL};
```
nit: missing spaces between elements (not your fault, but ...)

pentschev committed:
DOC: fix versionadded for *_like() shape argument

pentschev committed:
STY: add missing spaces in array initializers

pentschev commented:
Everything fixed, should be ready again for review.

mattip commented:
any more reviews for this, or is it ready to be merged?

pentschev commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1238,8 +1241,8 @@ PyArray_NewLikeArrayWithShape(PyArrayObject *prototype, NPY_ORDER order,
-    /* If it's not KEEPORDER, this is simple */
-    if (order != NPY_KEEPORDER) {
+    /* If it's not KEEPORDER, or there is a shape change, this is simple */
+    if (order != NPY_KEEPORDER || new_shape) {
```
But passing shape alone doesn't provide any information of strides, it will not be more than an assumption that keeping strides unchanged is important. Plus, we could only keep the strides for the dimensions that don't change.

eric-wieser replied:
The stride information comes from the source. The purpose of `KEEPORDER` is to preserve `np.argsort(arr.strides)`.

I think that the behavior should be:
* Same `ndim` - respect the stride ordering through keeporder
* different `ndim` - reject keeporder since it is meaningless, throw an exception asking the use to use corder or Forder.

pentschev replied:
But even if `ndim` remains the same, suppose source `shape=(5, 2)` and new `shape=(10000, 10)`, in such a case we can't respect strides. Another possibility is source `shape=(10000, 10)` and new `shape=(5, 2)`, we could respect the strides at the cost of huge memory waste, which is most likely going to be an unexpected behavior from the user.

eric-wieser replied:
keeporder means keep the relative order of the strides (`np.argsort(arr.strides)`), not the strides themselves (`arr.strides`).

pentschev replied:
Oh, I'm sorry, I got what you mean now. Yes, I agree with that, let me change that.

pentschev replied:
I think the new commits should cover this case now.

pentschev committed:
ENH: empty_like raises ValueError

This occurs when shape is defined and number of dimensions match but
order is 'K'.

pentschev committed:
TST: test for exception of *_like() functions

charris commented:
Should probably squash merge when the time comes.

charris commented:
Needs a release note, under `New features` I think.

pentschev committed:
DOC: release note for shape argument in *_like() functions

pentschev commented:
> Needs a release note, under `New features` I think.

Done.

eric-wieser commented on numpy/core/multiarray.py:
```diff
@@ -97,13 +97,22 @@ def empty_like(prototype, dtype=None, order=None, subok=None):
+    Raises
+    ------
+    ValueError
+        If len(ndim) different from a.ndim and order is 'K'
```
Should be `len(shape)`

pentschev replied:
Fixed.

eric-wieser commented on numpy/core/multiarray.py:
```diff
@@ -129,6 +138,9 @@ def empty_like(prototype, dtype=None, order=None, subok=None):
     """
+    if (shape is not None and prototype.ndim != len(shape) and
+            order not in 'CFA'):
+        raise ValueError("mismatching ndim can not keep order")
```
@shoyer: Is this the place for this type of check?

eric-wieser commented on numpy/core/tests/test_numeric.py:
```diff
@@ -2219,6 +2220,26 @@ def check_like_function(self, like_function, value, fill_value=False):
+
+                if d.ndim != len(s):
+                    assert_raises(ValueError, like_function, d, dtype=dtype,
+                                  shape=s, order='K', **fill_kwarg)
```
Needs an `else: assert_equal(np.argsort(like_function(...).strides) == np.argsory(d.strides)`

pentschev replied:
I just added this condition. Note we have also to check for non-C/F-layouts while raising an exception.

pentschev committed:
DOC: fix *_like() documentation on raises

pentschev committed:
BUG: *_like() raises for non-C/F-layout arrays

charris commented:
Looks like it is failing with `NPY_RELAXED_STRIDE_DEBUG=1`, not sure why.

pentschev commented:
@charris are they random, by any chance? I've tried to reproduce that error locally and wasn't able to.

charris commented:
@pentschev Looks like this in two places:
```
#if NPY_RELAXED_STRIDES_DEBUG
            /* For testing purpose only */
            if (dims[i] == 1) {
                strides[i] = NPY_MAX_INTP;
            }
#endif /* NPY_RELAXED_STRIDES_DEBUG */
```

charris commented:
Also, both `NPY_RELAXED_STRIDES_CHECKING` and `NPY_RELAXED_STRIDES_DEBUG` should be set to 1 in the environment. `NPY_RELAXED_STRIDES_CHECKING=1` is currently the default if it isn't in the environment.

charris commented:
Note that you need a clean build, they are both build time options. Hmm, I wonder if they are in the runtests.py build environment? You may need to do a local install, or modify the `setup.py` file. Modifying `setup.py` is probably the easiest thing.

pentschev commented:
@charris ah yes, now I can reproduce this. At first I thought they were just runtime variables. It worked by cleaning my build directory, export the variables and building again, I didn't need to modify `setup.py` at all.

pentschev commented:
Actually, I'm not really sure what to do with this error. It seems that strides for dimensions equal 1 gets set to `NPY_MAX_INTP` when `NPY_RELAXED_STRIDES_DEBUG=1` and strides is not specified in the C implementation.

The options I see are:

1. Ignore the assertion if any of the strides is equal to `NPY_MAX_INTP`; or
1. Meddle the test to avoid this condition getting tested.

In fact, the original shapes tested don't include dims equal 1. Is that on purpose for this exact reason?

Any ideas?

pentschev commented:
> In fact, the original shapes tested don't include dims equal 1. Is that on purpose for this exact reason?

It seems so, adding a test to `self.data` with dimension equal 1 fails too. So I'll for now change the shapes to avoid dimension 1 to prevent that failure, which seems to have been done before as well.

pentschev committed:
TST: change *_like() shapes to prevent NPY_RELAXED_STRIDE_DEBUG=1 failure

charris commented:
The message is that the stride values for dimensions that are 1 should not be depended on, so the test should not depend on them, they should be ignored. You can find an explanation in `doc/release/1.8.0-notes.rst`.

pentschev commented:
> The message is that the stride values for dimensions that are 1 should not be depended on, so the test should not depend on them, they should be ignored. You can find an explanation in `doc/release/1.8.0-notes.rst`.

So you're saying you agree with my change to prevent  dimension 1 in new shapes or that I should ignore assertion in those cases? I think both mean basically the same anyway.

eric-wieser commented:
Sounds like instead of testing argsort(strides), it should be testing the order of strides only for the axes not of length 1

charris commented:
> So you're saying you agree with my change to prevent dimension 1 in new shapes 

I think that is fine. If you think the test for dimension == 1 is important, you could mask out the stride. What matters is that you think the test is adequate to prove the implementation is good.

pentschev commented:
> Sounds like instead of testing argsort(strides), it should be testing the order of strides only for the axes not of length 1

I think the tests with shapes that don't include axes of length 1 are sufficient to test that. Would you agree with that?

eric-wieser commented:
Also fine

pentschev commented:
Ok, great. Everything passed now. It's ready for another review.

pentschev commented on numpy/core/multiarray.py:
```diff
@@ -71,9 +71,9 @@
+def empty_like(prototype, dtype=None, order=None, subok=None, shape=None):
     """
-    empty_like(prototype, dtype=None, order='K', subok=True)
+    empty_like(prototype, dtype=None, order='K', subok=True, shape=None)
```
I just noticed order and subok both default to None rather than the ('K' and True) values documentation proposes and other _like() functions have. Which one is the correct? It seems to me that the function definition should be order='K' and subok=True, am I correct?

shoyer replied:
It's a little confusing, but the arguments for this function (the dispatcher) need to default to None. We actually have a test that verifies this.

pentschev replied:
Should we then update the documentation to reflect the real defaults?

shoyer commented on numpy/core/multiarray.py:
```diff
@@ -129,6 +139,11 @@ def empty_like(prototype, dtype=None, order=None, subok=None):
            [  4.38791518e-305,  -2.00000715e+000,   4.17269252e-309]])
 
     """
+    if (shape is not None and order not in 'CFA' and
```
I think we need to do this check on the C side.

The function here is called to find arguments to dispatch on, e.g., if you write `np.ones_like(dask_array)`. But we don't want this method to fail on duck array types like dask just because they don't define some NumPy attributes.

pentschev replied:
Alright, I've moved the check to the C side now.

pentschev committed:
Move empty_like() exception to C implementation

mattip commented:
@shoyer, @eric-wieser: anything else needed here other than a tweak to the release notes to resolve a conflict?

hameerabbasi commented on numpy/core/numeric.py:
```diff
@@ -166,7 +176,7 @@ def zeros_like(a, dtype=None, order='K', subok=True):
+    res = empty_like(a, dtype=dtype, order=order, subok=subok, shape=shape)
     # needed instead of a 0 to get same result as zeros for for string dtypes
     z = zeros(1, dtype=res.dtype)
     multiarray.copyto(res, z, casting='unsafe')
```
Is the `casting=unsafe` needed, given that this is already the same `dtype`?

pentschev replied:
I honestly have no idea, this isn't part of this PR and I'd rather not change things out of its context.

hameerabbasi replied:
Fair enough. 👍

hameerabbasi commented on numpy/core/numeric.py:
```diff
@@ -286,7 +306,7 @@ def ones_like(a, dtype=None, order='K', subok=True):
     """
-    res = empty_like(a, dtype=dtype, order=order, subok=subok)
+    res = empty_like(a, dtype=dtype, order=order, subok=subok, shape=shape)
     multiarray.copyto(res, 1, casting='unsafe')
```
Ditto.

hameerabbasi commented on numpy/core/numeric.py:
```diff
@@ -395,7 +425,7 @@ def full_like(a, fill_value, dtype=None, order='K', subok=True):
     """
-    res = empty_like(a, dtype=dtype, order=order, subok=subok)
+    res = empty_like(a, dtype=dtype, order=order, subok=subok, shape=shape)
     multiarray.copyto(res, fill_value, casting='unsafe')
```
Ditto.

hameerabbasi reviewed (approved):
Some small comments. Overall, this LGTM.

pentschev committed:
Merge branch 'master' into add-shape-to-like-array-creation

pentschev commented:
> @shoyer, @eric-wieser: anything else needed here other than a tweak to the release notes to resolve a conflict?

Thanks for pointing out there was a merge conflict, I didn't see it before, but should be fixed now.

mrocklin commented:
Is there anything further to do here?

shoyer reviewed (approved):
This looks good to me!

charris commented on numpy/core/numeric.py:
```diff
@@ -127,12 +127,22 @@ def zeros_like(a, dtype=None, order='K', subok=True):
+    ------
+    ValueError
+        If len(shape) different from a.ndim and order is 'K', or
+        a is not a C/F-layout array
```
What does `C/F-layout` mean here?

charris replied:
I assume you mean the `order` argument has a invalid value? If so, maybe
```
        If `order` has an invalid value or `order` is 'K' and ``len(shape)``
        is different from ``a.ndim``.
```

pentschev replied:
I think what you mean with your suggestion is that `order` could be any random value different than `CFAK`, right? If that's the case, that's not what I my intent with the documentation, what I mean is that there's no particular memory layout, maybe I got the terminology wrong?

pentschev replied:
And by _no particular memory layout_, I mean something like the result of advanced indexing, for example, stealing that terminology from [indexing documentation](https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html#detailed-notes):

_The memory layout of an advanced indexing result is optimized for each indexing operation and no particular memory order can be assumed._

charris replied:
OK, I see what you mean. That is difficult to express exactly, maybe just brute force it: 
```
        If `order` is 'K'  and any of the following is true

            - ``len(shape) != a.ndim``,
            - `a` is neither C nor F contiguous.
```

charris replied:
Note that strictly speaking, it should be possible to add 1's to the shape without affecting the contiguity.
```
In [1]: a = ones((2,2), order='C')

In [2]: a.flags.c_contiguous
Out[2]: True

In [3]: a[None,:,:].flags.c_contiguous
Out[3]: True

In [4]: a = ones((2,2), order='F')

In [5]: a.flags.f_contiguous
Out[5]: True

In [6]: a[None,:,:].flags.f_contiguous
Out[6]: True
```
So possibly the correct check would be that the number of number of dimensions > 1 must match. @seberg Thoughts?

pentschev replied:
You're right, I can do that modification if necessary, but I'm not sure this is of utmost importance. Particularly, for our use cases I don't think that affects us.

pentschev committed:
Update *_like() ValueError documentation

pentschev commented:
I think the ci/circleci failure isn't related to this PR, could someone confirm that?

charris commented:
Looks unrelated to me also.

charris commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1183,28 +1183,43 @@ PyArray_NewFromDescrAndBase(
+        dims = PyArray_DIMS(prototype);
+        new_shape = 0;
+    }
+    else if (order == NPY_KEEPORDER && (ndim != PyArray_NDIM(prototype) ||
```
This breaks previous behavior which did not require the prototype to be contiguous.
```
In [1]: a = ones((3,4))[:,::2]

In [2]: a.flags
Out[2]: 
  C_CONTIGUOUS : False
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False

In [3]: ones_like(a, order='K')
Out[3]: 
array([[1., 1.],
       [1., 1.],
       [1., 1.]])
```
I'm not sure why an error should be raised, why not just override the shape and go with C order when a new shape is specified?

pentschev replied:
@charris what you're saying is like my original implementation. I modified this based on @eric-wieser's request here https://github.com/numpy/numpy/pull/13046#discussion_r260673995. For me it's fine either way.

eric-wieser replied:
@charris: This shouldn't break any existing code (assuming my comment below is addressed).

The thinking is that requesting the order (of the strides) to be kept is nonsensical if the the number of strides is not the same.

mattip commented:
CircleCI does not merge-then-build. The failure is due to #13231 missing from the PR. You could rebase off master or merge master into your branch.

pentschev committed:
Merge remote-tracking branch 'upstream/master' into add-shape-to-like-array-creation

pentschev commented:
@mattip thanks for the information, I merged master into this branch now.

eric-wieser commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1183,28 +1183,43 @@ PyArray_NewFromDescrAndBase(
+    else if (order == NPY_KEEPORDER && (ndim != PyArray_NDIM(prototype) ||
+                !(PyArray_IS_C_CONTIGUOUS(prototype) || PyArray_IS_F_CONTIGUOUS(prototype)))) {
+        PyErr_SetString(PyExc_ValueError,
+                        "mismatching ndim or non-C/F-layout arrays can not keep order");
```
Why are you checking `PyArray_IS_C_CONTIGUOUS` and `PyArray_IS_F_CONTIGUOUS` here?

pentschev replied:
Because we can't keep the order when there's _no particular memory layout_, just like in the discussion here:
https://github.com/numpy/numpy/pull/13046#discussion_r272338815

charris replied:
My current understanding of 'K' is that we try to keep the memory order _if possible_, not in all circumstances. Indeed, the former code fell back on recomputing the strides when the prototype was neither c/f continguous.

seberg replied:
First of all, we should make sure that this does not switch to an error suddenly. I agree that in most cases `K` probably means *if possible*.

I think the default may actually be `K` and not something like `None` (to mean `K if possible`), so that `K` is the most common option? If `K` was passed explicitly, there might be an argument.

There are some strides where it would be unambiguous, but I am not sure it is worth to bother about those.

pentschev replied:
Unless I'm miserably misunderstanding some of the comments, I think that there are two clashing opinions:

1. Keep order _if_ possible (as my original implementation did), otherwise, fallback to something like C-contiguous (currently backed by @charris and @seberg);
1. Raise an exception if the order is `K` but not possible to maintain it (backed by @eric-wieser, previously discussed in https://github.com/numpy/numpy/pull/13046#discussion_r260673995).

Feel free to correct me if I'm really misunderstanding it, otherwise, it would be nice to have here @charris, @eric-wieser and @seberg commenting here to reach an agreement before I do any further changes regarding this matter.

seberg replied:
@pentschev yes, but sorry, I was a bit confused. I think the important thing is that:
```
np.ones_like(np.ones((10, 10, 10), order="F")[::2, :, ::2]).flags
```
does not raise an error, but since the shape is not changed it is OK in any case. So, I am actually OK with just giving an error, could go either way though.

pentschev replied:
@seberg thanks for clarifying your concern. For my use case, both solutions are fine as well. At this point, I would just like to have a consensus to avoid modifying the PR back and forth.

As soon as there's a confirmation from both @charris and @eric-wieser we can move forward with the PR, either to merge it or to make changes as necessary.

pentschev replied:
Just one more thing I thought about checking now, the [documentation on `numpy.array`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html) doesn't mention any errors/exceptions for order `K, saying only:

_F & C order preserved, otherwise most similar order_

So if there's an interest in maintaining the API closer to what happens in `numpy.array`, maybe raising an exception from these `*_like()` functions could be counterintuitive.

eric-wieser replied:
> Because we can't keep the order when there's no particular memory layout,

My understanding is that we can and do - that's what the "sorted stride perm" function does

pentschev replied:
My understanding from [this comment](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/shape.c#L814) is that it will do that provided that input arrays are C or F contiguous. For the condition we are talking about here, the array is neither. In other words, what does it mean to keep the order if it's not known?

eric-wieser replied:
It means keep `argsort(strides[shape !=1])` unchanged (ignoring ties)

eric-wieser replied:
I think I need to play around with that function, my interpretation might be wrong

pentschev replied:
@eric-wieser could you let me know about your findings then? I'm sorry, but I don't understand what you mean with your previous statement, it may be getting a little too complex for my understand of NumPy handling of strides and orders.

eric-wieser replied:
Sure, here's the case I'm thinking of:
```python
>>> a = np.zeros((20, 30, 40)).transpose((1, 0, 2))[::10,::10,::10]

>>> a.strides
(3200, 96000, 80)

>>> np.copy(a, order='K').strides
(32, 96, 8)
```

note how the order of the strides is kept.

This is the behavior I would expect for `empty_like` too

pentschev replied:
@eric-wieser I believe the new commit I pushed covers now the case you're looking for. Extending the case from your previous comment, this is what we get with this new commit:

```python
>>> np.empty_like(a, shape=(3, 2, 4), order='K').strides
(32, 96, 8)
```

Can you confirm this is what you had in mind?

eric-wieser replied:
That behavior looks correct to me now. Well take one more look at this tonight.

pentschev replied:
@eric-wieser just checking, have you had the chance to have a look at this? Do you think we're good to merge this now?

eric-wieser reviewed (dismissed):
Drawing attention to my lack of approval here - we should not be special casing `PyArray_IS_C_CONTIGUOUS` and `PyArray_IS_F_CONTIGUOUS`

eric-wieser commented:
@seberg:

> I agree that in most cases K probably means if possible.
>
> I think the default may actually be K and not something like None (to mean K if possible), so that K is the most common option? If K was passed explicitly, there might be an argument.

This touches on the `np.never_copy` thing that came up before and sadly seems to have stalled in a naming argument. If there's no precedent yet for `order='K'` meaning _"`'K'` if possible"_, then this is our chance to not make the same mistake we did with copy. If you can find precedent for that meaning already, then I agree with you.

If not, then I'd perhaps be inclined to use `order=None` to mean `K if possible else C` or something

pentschev committed:
Rearrange stride computation for *_like() if new shape and order='K'

eric-wieser commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1232,12 +1246,12 @@ PyArray_NewLikeArray(PyArrayObject *prototype, NPY_ORDER order,
-    /* If it's not KEEPORDER, this is simple */
-    if (order != NPY_KEEPORDER) {
+    /* If it's not KEEPORDER, or there is a shape change, this is simple */
+    if (order != NPY_KEEPORDER || (new_shape && ndim != PyArray_NDIM(prototype))) {
```
You can simplify this back to what it was before now that `NPY_KEEPORDER` implies no shape change, right?

eric-wieser commented:
@pentschev: My one worry here is that you probably would like `np.empty_like(np.zeros(2), shape=(2,2))` to work, but with the change I was asking for it requires you to pass an explicit order.

My comment above of:

> If not, then I'd perhaps be inclined to use `order=None` to mean `K if possible else C` or something

probably applies here

pentschev commented:
> @pentschev: My one worry here is that you probably would like `np.empty_like(np.zeros(2), shape=(2,2))` to work, but with the change I was asking for it requires you to pass an explicit order.
> 
> My comment above of:
> 
> > If not, then I'd perhaps be inclined to use `order=None` to mean `K if possible else C` or something
> 
> probably applies here

I'm not sure this is a good idea, the only function that has `order=None` by default is `empty_like`, all others have `order=K`. Wouldn't that be misleading since they could have apparent different behaviors?

pentschev commented:
Ok, looks like `order=None` means the same as `order=K` on the C side of things. So your suggestion is to get rid of raising a `ValueError` and just implying `order=C` when there's a change of dims and order=`K`, am I right @eric-wieser?

mrocklin commented:
Checking in here.  Is there anything we can do to accelerate this?

pentschev committed:
Change handling of order= for *_like()

- If order='K' try to keep, otherwise, order='C' is implied
- Do not raise ValueError anymore

pentschev committed:
Merge branch 'master' into add-shape-to-like-array-creation

pentschev commented:
I pushed some changes which I believe now handle @eric-wieser's latest request.

eric-wieser commented:
What I was hoping for is a distinction between order=none and order=k at the c level, and some feedback from other maintainers on whether that's a good idea.

pentschev commented:
> What I was hoping for is a distinction between order=none and order=k at the c level, and some feedback from other maintainers on whether that's a good idea.

I guess that would probably need a change of signature of either `empty_like()` (currently `order=None`) or all the other `_like()` functions (currently `order='K'`), which probably wouldn't be something that everyone would be keen to have done here. Plus, the change you're suggesting feels to me much broader and likely to have a potentially larger impact than what we intend with this PR.

pentschev committed:
Fix *_like() tests

pentschev commented:
It seems to me that the failures are unrelated to this PR. Could anyone confirm this?

mattip commented:
@eric-wieser could we merge this PR as-is and open an issue to fix `order` in `empty_like`, or are there implications that prevent merging this?

stefanv commented:
This has been approved by two core contributors, and seem to address @eric-wieser's major concerns.  Eric, do you want to go ahead and merge?

eric-wieser commented:
I started writing a justification for changing the default to `None`, but then found that we document `K` as
```
/* An order as close to the inputs as possible */
```

Given this, I think we should do one of three things for when `len(shape) != ndim`
1. Try harder to be as close as possible before merging this patch
   1. Interpret order=`'K'` to mean "keep the order of `out.strides[:m]` compared to `in_.strides[:m]` where `m=min(ndim, len(shape))`
   2. Interpret order=`'K'` to mean "keep the order of `out.strides[-m:]` compared to `in_.strides[-m:]` where `m=min(ndim, len(shape))` (consistent with broadcasting)
2. Accept this patch as is, declare 1. as a backward-compatible change, and schedule it at some point in future without needing any kind of deprecation cycle
3. Accept this patch, declare 1. as a backwards compatible and never fix it

My preference here is 1.ii, but really my goal is to ensure we don't pick 3

eric-wieser commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1259,14 +1267,14 @@ PyArray_NewLikeArray(PyArrayObject *prototype, NPY_ORDER order,
             strides[i_perm] = stride;
-            stride *= shape[i_perm];
+            stride *= dims[i_perm];
         }
```
My 1.ii proposal would look like changing everything between L1261 and here to:
```C
if (PyArray_NDIM(prototype) >= ndim) {
    int leading_dims = PyArray_NDIM(prototype) - ndim;

    /* Use only the trailing strides */
    PyArray_CreateSortedStridePerm(ndim,
                                   PyArray_STRIDES(prototype) + leading_dims,
                                   strideperm);
    /* Build the new strides */
    stride = dtype->elsize;
    for (idim = ndim-1; idim >= 0; --idim) {
        npy_intp i_perm = strideperm[idim].perm;
        strides[i_perm] = stride;
        stride *= shape[i_perm];
    }
}
else {
    int leading_dims = ndim - PyArray_NDIM(prototype);
    /* Use all the strides */
    PyArray_CreateSortedStridePerm(PyArray_NDIM(prototype),
                                   PyArray_STRIDES(prototype),
                                   strideperm);

    /* Build the new trailing strides */
    stride = dtype->elsize;
    for (idim = PyArray_NDIM(prototype)-1; idim >= 0; --idim) {
        npy_intp i_perm = strideperm[idim].perm + leading_dims;
        strides[i_perm] = stride;
        stride *= shape[i_perm];
    }

    /* Create remaining leading strides as C order */
    for (idim = leading_dims; idim >= 0; --idim) {
        strides[idim] = stride;
        stride *= shape[idim];
    }
}
```

eric-wieser replied:
Or trading branching for state:
```C
int leading_src_dims = 0;
int leading_dst_dims = 0;
int shared_dims;

if (PyArray_NDIM(prototype) >= ndim) {
    shared_dims = ndim;
    leading_src_dims = PyArray_NDIM(prototype) - ndim;
}
else {
    shared_dims = PyArray_NDIM(prototype);
    leading_dst_dims = ndim - PyArray_NDIM(prototype);
}

/* Use only the trailing strides from the source */
PyArray_CreateSortedStridePerm(shared_dims,
                               PyArray_STRIDES(prototype) + leading_src_dims,
                               strideperm);

/* Build the destrination trailing strides */
stride = dtype->elsize;
for (idim = ndim-1; idim >= 0; --idim) {
    npy_intp i_perm = strideperm[idim].perm + leading_dst_dims;
    strides[i_perm] = stride;
    stride *= shape[i_perm];
}

/* Create remaining leading strides as C order */
for (idim = leading_dst_dims; idim >= 0; --idim) {
    strides[idim] = stride;
    stride *= shape[idim];
}
```

eric-wieser commented:
An example of what I'd expect from the behavior of 1ii is, I think:
```python
# note: number of * shows stride order
a = np.zeros((2, 3, 5, 7), dtype=np.uint8, order='F')
assert a.strides == (1, 1*2, 1*2*3, 1*2*3*5)
a_t = a.transpose((2, 0, 3, 1))
assert a_t.strides == (1*2*3, 1, 1*2*3*5, 1*2)

b = np.zeros_like(a, shape=a.shape[1:], dtype=a.dtype)
assert b.strides == (1, 1*3, 1*3*5)  # note: just remove the 2 factors
b_t = np.zeros_like(a_t, shape=a_t.shape[1:], dtype=a_t.dtype)
assert b.strides == (1, 1*2*3, 1*2)  # note: just remove the 5 factors

c = np.zeros_like(a, shape=(11,) + a.shape, dtype=a.dtype)
assert c.strides == (1*2*3*5*7,) + a.strides  # C order beyond matching dimensions
c_t = np.zeros_like(a_t, shape=(11,) + a_t.shape, dtype=a_t.dtype)
assert c.strides == (1*2*3*5*7,) + a.strides
```

rgommers commented:
> Accept this patch as is, declare 1. as a backward-compatible change, and schedule it at some point in future without needing any kind of deprecation cycle

I'm fine with this - `order='K'` basically has always meant "we try our best, and in practice it just means you get C or F, that's all". 

These very detailed considerations on strides have not been considered before AFAIK in relation to the `order` keyword, and if external code really depends on it then I'd consider them to rely on implementation details.

I would recommend merging this right now, and approving a future change without deprecation cycle. Requiring changes almost *ad infinitum* before merging is potentially discouraging for contributors, especially if there's back and forth between maintainers who don't agree on small details.

eric-wieser commented:
> we try our best, and in practice it just means you get C or F, that's all"

I think we've actually managed to do a good job at doing better than picking between C/F

> if external code really depends on it 

To be clear, I'm suggesting a performance dependency, not a behavior dependency.

> I would recommend merging this right now, and approving a future change without deprecation cycle.

Sounds good to me

seberg commented:
Great, thanks guys. Yes, merging as is good, I would have been fine with the error path for now even. Should we discuss about where to go?

I would be on board for the 1.ii. choice (undefined part is C-order, except when all is F-order already I suppose). There is maybe some question what to do if the array is almost F – i.e. F-order, but not contiguous). It seems the closest we will get to "keeping" some layout. In general, frankly, I doubt it will be used much (with someone caring about the result) anyway though.

pentschev commented:
Thanks so much everyone for the reviews!

FWIW, I've tried testing the code you suggested "as-is" @eric-wieser, it didn't seem to work for all cases. I admit I don't really understand all this complex striding though, so maybe I did something wrong. However, feel free to ping me if I can help with any of the fixes.

rgommers commented:
Thanks for the contribution @pentschev!
